ojdbc-provider-oci/src/main/java/oracle/jdbc/provider/oci/resource/VaultConnectionStringProvider.java

```
@@ -0,0 +1,139 @@
+  };
+
+  /**
+   * A public no-arg constructor used by {@link java.util.ServiceLoader} to
+   * construct an instance of this provider.
+   */
+  public VaultConnectionStringProvider() {
+    super("vault-tnsnames", PARAMETERS);
+  }
+
+  /**
+   * Retrieves a database connection string from the tnsnames.ora file stored
+   * in OCI Vault.
+   * <p>
+   * This method accesses the file in OCI Vault, decodes it if it is
+   * base64-encoded, and parses the `tnsnames.ora` content. The method
+   * returns the connection string associated with the specified alias from
+   * the `tnsnames.ora` file.
+   * The file can either be stored as plain text or as base64-encoded content
+   * in the vault.
+   * </p>
+   *
+   * @param parameterValues The parameters required to access the `tnsnames.ora`
+   * file in OCI Vault, including the vault OCID and the tns-alias.
+   * @return The connection string associated with the specified alias
+   * in the `tnsnames.ora` file.
+   * @throws IllegalStateException If there is an error reading the `tnsnames.ora`
+   * file, decoding its content, or accessing the OCI Vault.
+   * @throws IllegalArgumentException If the specified alias is invalid or
+   * does not exist in the `tnsnames.ora` file.
+   */
+  @Override
+  public String getConnectionString(Map<Parameter, CharSequence> parameterValues) {
+
+    // Retrieve the secret containing tnsnames.ora content from OCI Vault
+    String secretValue = getVaultSecret(parameterValues)
+                           .getBase64Secret();
+
+    byte[] fileBytes = Base64.getDecoder().decode(secretValue);
+
+    TNSNames tnsNames;
+    try (InputStream inputStream = new ByteArrayInputStream(fileBytes)) {
+      tnsNames = TNSNames.read(inputStream);
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to read tnsnames.ora content", e);
+    }
+
+    String alias;
+    try {
+      alias = parseParameterValues(parameterValues).getRequired(TNS_ALIAS);
```

Shouldn't you check if the parameter is present sooner, so that it if parameter is missing you can fail faster, before getting the secret from OCI and parsing the file?